models/core-java/src/main/java/org/commonjava/indy/model/core/Group.java

```diff
@@ -86,17 +88,28 @@ public boolean addConstituent( final StoreKey repository )
 
         if ( constituents == null )
         {
-            constituents = new ArrayList<>();
+            constituents = new LinkedList<>();
         }
 
-        synchronized ( constituents )
+        synchronized ( monitor )
         {
             if ( constituents.contains( repository ) )
             {
                 return false;
             }
 
-            return constituents.add( repository );
+            // We will add new hosted repo as the first member in group, as this hosted will most likely
+            // contain the most recent dependencies for the subsequent build.
+            if ( repository.getType() == StoreType.hosted )
+            {
+                ( (LinkedList<StoreKey>) constituents ).push( repository );
```

Comment:
Is LinkedList.push(..) better than ArrayList.add(0,..)? Does LinkedList have performance implications for iterating through the items in the list?

Reply:
Maybe ArrayList is still more suitable here as the most often used cases are reading than writing. I'll change it back.